pageserver/src/tenant/timeline.rs

```diff
@@ -770,17 +946,43 @@ impl Timeline {
                 "initial size calculation",
                 false,
                 async move {
-                    let calculated_size = self_clone.calculate_logical_size(init_lsn)?;
-                    let result = spawn_blocking(move || {
-                        self_clone.current_logical_size.initial_logical_size.set(calculated_size)
-                    }).await?;
-                    match result {
-                        Ok(()) => info!("Successfully calculated initial logical size"),
-                        Err(existing_size) => error!("Tried to update initial timeline size value to {calculated_size}, but the size was already set to {existing_size}, not changing"),
+                    let mut timeline_state_updates = self_clone.subscribe_for_state_updates();
+                    tokio::select! {
+                        calculation_result = async {
+                            let calculated_size = self_clone.calculate_logical_size(init_lsn)?;
+                            let result = spawn_blocking(move || {
+                                self_clone.current_logical_size.initial_logical_size.set(calculated_size)
+                            }).await?;
```

Comment:
I don't see anything blocking here. Why the `spawn_blocking`?

Reply:
Shouldn't the `calculate_logical_size` be within this block, which would then make sense?

I think the cancellation by `tokio::select` would then become harmless, however it would never be actually be cancelled because it would already block the runtime thread polling this task. To cancel the blocking `calculate_logical_size` I think we need to share some sort `AtomicBool` or better to actually cancel it within `spawn_blocking`.

Reply:
Good catch, somehow `calculate_logical_size` slipped away.
But I don't see anything blocking inside that either (yet it's vast, I could've missed things) so I wonder, if I could remove `spawn_blocking` entirely.
I think I won't be able to push a cancellation into proper parts of the `calculate_logical_size`, that's why I wonder.

Reply:
I was thinking maybe check any given `AtomicBool` or `CancellationToken` within the inner loop: https://github.com/neondatabase/neon/blob/0e998aa84bd003481af80e1c4d8c066868368da2/pageserver/src/pgdatadir_mapping.rs#L387

But it's possible I am looking at the wrong place.

Reply:
Unfortunately, we cannot use `CancellationToken` in `pgdatadir_mapping` since it has zero `async` and some other code is blocking (I've found that `page_cache` is), so I went with a safer `Arc<AtomicBool>` approach, even though it's not very fancy.